src/Fmt_ast.ml

```diff
@@ -1696,6 +1699,36 @@ and fmt_expression c ?(box = true) ?epi ?eol ?parens ?(indent_wrap = 0) ?ext
                           ( if c.conf.indicate_multiline_delimiters then " )"
                           else ")" ) )
                     $ fmt_if (not last) "@ "
+                | `Fit_or_vertical ->
+                    let pro =
+                      fmt_if_k
+                        (not
+                           (Location.is_single_line pexp_loc c.conf.margin))
+                        (break_unless_newline 1000 2)
+                    in
+                    hovbox 0
+                      ( ( match xcnd with
+                        | Some xcnd ->
+                            hvbox
+                              (if parens then -2 else 0)
+                              ( hvbox 2
+                                  ( fmt_if (not first) "else "
+                                  $ fmt "if"
+                                  $ fmt_if_k first
+                                      (fmt_extension_suffix c ext)
+                                  $ fmt_attributes c ~pre:(fmt " ") ~key:"@"
+                                      pexp_attributes
+                                  $ fmt "@ " $ fmt_expression c xcnd )
+                              $ fmt_or parens "@;<1 2>" "@ "
+                              $ fmt "then" )
+                        | None -> fmt "else" )
+                      $ fmt_if parens_bch " (" $ fmt "@;<1 2>"
+                      $ fmt_expression c ~pro ~eol:(fmt "@;<1 2>")
+                          ~box:false ~parens:false xbch
+                      $ fmt_if parens_bch
+                          ( if c.conf.indicate_multiline_delimiters then " )"
+                          else ")" ) )
+                    $ fmt_if (not last) "@ "
```

Comment:
This function is extremely big. Also, I see a bit of repetition, would the `Params` thing help ?

Reply:
It will once we find a common box structure for a the cases. I didn't look at that, maybe it could take some time, I think this factorization step could be done in a future PR, since it will also impact the existing cases and the diff would be less readable than it is now.